core/jobs/batch_jobs/remove_profile_picture_data_url_field_jobs.py

```diff
@@ -0,0 +1,77 @@
+# coding: utf-8
+#
+# Copyright 2021 The Oppia Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Remove profile_picture_data_url field from UserSettingsModel."""
+
+from __future__ import annotations
+
+from core.jobs import base_jobs
+from core.jobs.io import ndb_io
+from core.jobs.transforms import job_result_transforms
+from core.jobs.types import job_run_result
+from core.platform import models
+
+import apache_beam as beam
+
+MYPY = False
+if MYPY:  # pragma: no cover
+    from mypy_imports import user_models
+
+(user_models,) = models.Registry.import_models([models.Names.USER])
+
+
+class RemoveProfilePictureFieldJob(base_jobs.JobBase):
+    """Remove profile_picture_data_url from UserSettingsModel."""
+
+    def _remove_profile_field(
+        self, user_model: user_models.UserSettingsModel
+    ) -> user_models.UserSettingsModel:
+        """Remove profile_picture_data_url field from the model.
+
+        Args:
+            user_model: UserSettingsModel. The user settings model.
+
+        Returns:
+            user_model: UserSettingsModel. The updated user settings model.
+        """
+        if 'profile_picture_data_url' in user_model._properties:  # pylint: disable=protected-access
```

Comment:
I don't think I understand the launch process for this PR.

In what order will you do the following:

- Merge this PR into develop
- Deploy the PR to the prod server
- Run the job

If it is the above order, then what does our datastore look like between deploying the PR to the prod server and running the job? Will the profile_picture_data_url field be accessible when the job fetches the model, if it is not in the ndb.Model class description? In what way have you tested this?

An alternative approach might be to set it to None for all models, and mark the field as deprecated. We can then remove the field at leisure.

Also, what happens if new models are created after running the job? Will their profile_picture_data_url field get populated? I think I need to see a really clear description of the launch process here, same as I asked about for M1.

Reply:
Hey @seanlip, thanks for the comment. Yes, your order is correct that is how we are planning to get things done. The complete launch plan for the profile_picture part is mentioned below:
1. AUDIT: Run `AuditInvalidProfilePictureJob` to get invalid profile picture data.

2. MIGRATION STEPS:
        a. Run `FixInvalidProfilePictureJob` in case the errors from above job reported.
        b. Run `AuditInvalidProfilePictureJob` to verify that the profile pictures are fixed.
        
3. STORE IMAGES TO GCS:
        a. Store images to GCS via job.
        b. Run an audit job to fetch the images via URL(user/<username>/profile_picture.webp).
        
4. MODIFICATIONS in FRONTEND and BACKEND:
        a. Modify the frontend to get the profile images directly from the GCS instead of getting the base64 string using this URL - 'user/<username>/profile_image.webp'.
        b. Modify the backend to upload images directly instead of storing the base64 string. This includes updating the handlers that send data to the frontend.
        c. Remove all the occurences of `profile_picture_data_url` except from the `UserSettingsModel`.
        
5. CLEANUP: Write a job to remove the `profile_picture_data_url` field from the `UserSettingsModel` and manually verify that the field is removed.

6. Run Step 3 again just to be sure if any user images were updated or added while the deployment.

> what does our datastore look like between deploying the PR to the prod server and running the job?
I'm a bit confused about this part please correct me if I'm wrong, while we are deploying the changes to the server don't we go in to the maintenance mode for a while because while deploying, the server will not be responsive. If that is the case then no new data will be coming till we are deploying also note that after the deployment is done and the new data comes, let's say the user updates or uploads the profile picture the data will directly be pushed to the GCS so we will not to worry about that part.

> Will the profile_picture_data_url field be accessible when the job fetches the model
I was talking about this part with @vojtechjelinek and he mentioned that it will work as we previously did the same part for `MapReduce` and it worked. Also we will be testing it on the backup server which will give us some more confirmation. And yes I will be adding more test to this part on the local part as Vojta mentioned above and can provide you further confirmation.

> what happens if new models are created after running the job
Please note that all the new data will directly be uploaded to the GCS so we will not need to worry about that part.

Please note that this PR is the last part in the process and will be merged in the last after everything.
Please let me know if any point is not clear and I will try to elaborate it further.
Thanks

Reply:
Oh sure, I will mention the steps according to the PR's open. Please note that the PRs mentioned below will be ran and merged in production in the given sequence only.
1. #16492 - Audit and fix invalid profile pictures in UserSettingsModel.
    a. Merge PR into develop.
    b. Deploy PR to production.
    c. Run `AuditInvalidProfilePictureJob` to audit invalid profile pictures.
    d. Run `FixInvalidProfilePictureJob` to fix the invalid profile pictures.
    e. Run `AuditInvalidProfilePictureJob` to check if all the profile pictures are fixed.

2. #16541 - Store and audit profile pictures to GCS.
    a. Merge PR into develop.
    b. Deploy PR to production.
    c. Run `StoreProfilePictureToGCSJob` to push the profile pictures to GCS.
    d. Run `AuditProfilePictureFromGCSJob` to check the images stored on GCS and the images in model are same.

3. #16521 - Backend and frontend changes to store profile pictures directly to GCS and to handle takeout and wipeout services.
    a. Merge PR into develop.
    b. Deploy PR to production.
    From now on the images will be directly deployed to GCS.

4. #16584 - This PR: Will remove all the occurrences of `profile_picture_data_url` field from UserSettingsModel.
    a. Merge PR into develop.
    b. Deploy PR to production.
    c. Run `RemoveProfilePictureFieldJob`, this will remove the field from the UserSettingsModel.

5. Repeat step 2 to in case the UserSettingsModel get populated while we were deploying the changes to the server.

> Also, has the audit job been run on the backup server yet

We ran the jobs from step 1 on the backup server and onto which we were able to fix the invalid profile pictures but we found out that for some images gravatar was not generated and default value was assigned to them. We were using `[fetch_gravatar](https://github.com/oppia/oppia/blob/develop/core/domain/user_services.py#L377)` function for the generation. Now I and @vojtechjelinek are trying to discuss what is the root cause for this and why we are not able to generate the gravatar for some images though we discussed few things but not able to get to the exact reason.

> Note that we typically don't go into maintenance mode on the server, to avoid service downtime. If a maintenance mode window is needed, have you discussed this explicitly with Vojta and Kevin, the release coordinators, and have they agreed to it?

Yes, so I just had one small doubt which I cleared with @vojtechjelinek which was "What happens to the request made to the server when we are deploying some code at that time?" to which I get to know that the server will work according to the old code till the changes are deployed.

> I don't understand what you mean by "it will work as we previously did the same part for MapReduce and it worked". Could you point to the PR in question, or docs that confirm this? The other checks sound good.

I think @vojtechjelinek can give us more clarity on this part mainly the links to the PR and the changes because I only 
briefly discussed this part with him and he mentioned that it would work and we have to test it in order to be completely sure as we have also done the similar thing in past. Hey @vojtechjelinek can you please give us some more pointers here?

Please let me know if anything else is not clear or does not make sense.
Thanks

Reply:
Regarding the last paragraph here is an example of what we did in MapReduce https://github.com/oppia/oppia/blob/v3.0.4/core/domain/user_jobs_one_off.py#L444. It will hopefully work the same way in the Beam jobs.

It is fine when the model definition doesn't contain some field but the field is still stored in the actually model in the datastore, the model will still get loaded. This job should just make sure that even the models in the datastore do not have the field.

Reply:
Thanks @lkbhitesh07 and @vojtechjelinek.

Just to check, is the plan to make 4+ separate deployments to production in 4 separate releases? That seems a bit excessive so I wanted to check. If some of them will be batched then can you be specific about which ones are batched in a single release?

For the missing gravatars, are you talking about gravatars not being generated for new users, or that some historical data didn't have gravatars? If it's the latter then just use a default gravatar and don't worry about it, it's not worth investigating in detail.

Reply:
Right, I think we can batch all of them together in a single release as I do not see anything that will not work. I would like to confirm this part with @vojtechjelinek just to confirm.

> For the missing gravatars, are you talking about gravatars not being generated for new users, or that some historical data didn't have gravatars? If it's the latter then just use a default gravatar and don't worry about it, it's not worth investigating in detail.

It is the later one, thanks for the response. Also, currently the default image that are assigned do not have the correct dimensions, it has 76x76 and we want all the images to have 150x150, so we just have to replace the default images, rest all is done.
Please let me know if you have any other questions.
Thanks

Reply:
1, 2, 3 need to be done together, 4 would be better to be done in a later release.

Reply:
OK, thanks @vojtechjelinek. @lkbhitesh07, based on @vojtechjelinek's comment, could you please post an updated version of your launch plan here?

Reply:
Sure @seanlip, I have also added the part where we are testing on the backup server.
1. #16492 - Audit and fix invalid profile pictures in UserSettingsModel.
a. Deploy the PR to the backup server.
b. Run AuditInvalidProfilePictureJob to audit invalid profile pictures.
c. Run FixInvalidProfilePictureJob to fix the invalid profile pictures.
d. Run AuditInvalidProfilePictureJob to check if all the profile pictures are fixed.
e. After the successful run of all the jobs, merge the PR to develop.

2. #16541 - Store and audit profile pictures to GCS.
a. Deploy the PR to the backup server.
b. Run StoreProfilePictureToGCSJob to push the profile pictures to GCS.
c. Run AuditProfilePictureFromGCSJob to check the images stored on GCS and the images in model are same.
d. After the successful run of all the jobs, merge the PR to develop.

3. #16521 - Backend and frontend changes to store profile pictures directly to GCS and to handle takeout and wipeout services.
a. Merge PR into develop.
From now on the images will be directly deployed to GCS.

4. Cut the release and deploy 16492, 16541, and 16521 PRs to production.
a. Run AuditInvalidProfilePictureJob to audit invalid profile pictures.
b. Run FixInvalidProfilePictureJob to fix the invalid profile pictures.
c. Run AuditInvalidProfilePictureJob to check if all the profile pictures are fixed.
d. Run StoreProfilePictureToGCSJob to push the profile pictures to GCS.
e. Run AuditProfilePictureFromGCSJob to check the images stored on GCS and the images in model are same.

5. #16584 - This PR: Will remove all the occurrences of the profile_picture_data_url field from UserSettingsModel.
a. Deploy the PR to the backup server.
b. Run RemoveProfilePictureFieldJob, this will remove the field from the UserSettingsModel.
c. Run the jobs from step 2, which are StoreProfilePictureToGCSJob and AuditProfilePictureFromGCSJob.
d. After the successful run of all the jobs, merge the PR to develop.

6. Deploy 16584 to production.
a. Run RemoveProfilePictureFieldJob, this will remove the field from the UserSettingsModel.
b. Run the jobs from step 2, which are StoreProfilePictureToGCSJob and AuditProfilePictureFromGCSJob.

Please let me know if you have any more questions.
Thanks

Reply:
This looks excellent. Please provide proof of successful tests on the backup server before the corresponding PRs are approved, but we can refer to this as the official plan. Thanks!

Reply:
Actually, wait, I have one question about the 4th step.

Shouldn't we run the StoreProfilePictureToGCSJob **just before** deploying? Otherwise, once you deploy, the field will be removed -- wouldn't that make the previous profile picture inaccessible?

Also, could you confirm what happens to the older models when you run the StoreProfilePictureToGCSJob twice? Will it overwrite images that have already been updated in the interim?

Reply:
I don't think we need to run it before step 4, because the code in step 3 will already work with images in GCS so every new user will have the picture stored in GCS so there will be no data loss.

Reply:
I think @vojtechjelinek has replied to the first part of the question and I agree with that. Regarding the second part which is to run the `StoreProfilePictureToGCSJob` twice - I have submitted the testing request for #16497 where we are running the `TestGCSIoWriteJob` job twice to confirm that we do not get any error when we run the job again and can overwrite the data in the second run. [This](https://docs.google.com/document/d/1AEb-RhoctyU_Ymh0QlwONHkAD7bUva99PEF_iMIwzok/edit) sequence will be followed when we will test the features I added for GCS Io implementation. Please note that the `StoreProfilePictureToGCSJob` is based upon the GCS Io implementation which means that the `StoreProfilePictureToGCSJob` internally uses the GCS Io functionality that I introduced in the 16496.
Thanks

Reply:
Thanks @lkbhitesh07 -- I have looked at the [release instructions doc](https://docs.google.com/document/d/1AEb-RhoctyU_Ymh0QlwONHkAD7bUva99PEF_iMIwzok/edit#). The expected conditions are very unclear. Please fix these and leave a comment here when you have done so, so that I can take a look at it again -- currently I cannot process your previous comment properly because of the gaps here.

Regarding the first point, I think I understand what you are saying. Just to check -- can you and @vojtechjelinek confirm that after "16492, 16541, and 16521 PRs" are deployed to the server, then the profile_picture_data_url field in UserSettingsModel is no longer used at all?

Reply:
[Release instructions doc](https://docs.google.com/document/d/1AEb-RhoctyU_Ymh0QlwONHkAD7bUva99PEF_iMIwzok/edit#): I have updated the instructions, please let me know if there is something that is still unclear and I will update that ASAP.

Yes, I can confirm that after "16492, 16541, and 16521 PRs" are deployed to the server, the `profile_picture_data_url` field will no longer be used.
cc @vojtechjelinek

Reply:
Thanks, the release doc is much clearer now! I just had one last comment.

For the rest, sounds good, thanks!